### PR TITLE
Save df decorator

### PR DESCRIPTION
**Pull request by hadrilec**
This PR contains a decorator used to cache pandas dataframe, macrodata module functions use this decorator to to reduce the number of requests to api.insee.fr.
It should answer this issue #195

### CODE REVIEW

**Review by tfardet — CHANGES_REQUESTED** [review]
Looks pretty good, thanks!
Just some minor comments and I'll try to check it locally.

I added a suggestion about moving everything to parquet but if that's too much work it can definitely go in another PR.

File: pynsee/utils/save_df.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                day_lapse = (insee_date_time_now - file_date_last_modif).days
+
+                if day_lapse_max is not None:
+                    if day_lapse > day_lapse_max:
+                        update = True                    
+            
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", category=ShapelyDeprecationWarning)
+
+                if (not os.path.exists(file_name)) | (update is True):
```
Comment (tfardet) [inline pynsee/utils/save_df.py:79]:
```suggestion
                if (not os.path.exists(file_name)) or update:   
```

File: pynsee/utils/save_df.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            
+            if parquet:
+                file_name += ".parquet"
+            else:
+                file_name += ".pkl"            
+                                    
+            if any([(a == 'update') & (kwargs[a] == True) for a in kwargs.keys()]):
+                update = True
+            else:
+                update = False
```
Comment (tfardet) [inline pynsee/utils/save_df.py:50]:
```suggestion
            update = any((a == 'update') and (kwargs[a] is True) for a in kwargs)
```

File: pynsee/utils/save_df.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                                    
+            if any([(a == 'update') & (kwargs[a] == True) for a in kwargs.keys()]):
+                update = True
+            else:
+                update = False  
+
+            if any([(a == 'silent') & (kwargs[a] == True) for a in kwargs.keys()]):
+                silent = True
+            else:
+                silent = False
```
Comment (tfardet) [inline pynsee/utils/save_df.py:55]:
```suggestion
            silent = any((a == 'silent') and (kwargs[a] is True) for a in kwargs)
```

**Review by tfardet — CHANGES_REQUESTED** [review]

File: pynsee/utils/save_df.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            
+            data_folder = _create_insee_folder()
+            string_file_arg = [str(kwargs[a]) for a in kwargs.keys() if a not in ['update', 'silent']]
+            string_file_arg += [func.__name__] + [str(a) for a in args]            
+            file_name = data_folder + "/" + _hash(''.join(string_file_arg)) 
+            
+            if parquet:
+                file_name += ".parquet"
+            else:
+                file_name += ".pkl"
```
[the remark this answers is not in the document]
Comment (hadrilec) [reply pynsee/utils/save_df.py:45]: as long as we use the save_df decorator everywhere it should not be too much cumbersome to move entirely to parquet

File: pynsee/macrodata/get_dataset_list.py [hunk 5]
```diff
@@ -1,21 +1,23 @@
...
 from pynsee.utils._get_temp_dir import _get_temp_dir
+from pynsee.utils.save_df import save_df
 
 import logging
 logger = logging.getLogger(__name__)
 
-@lru_cache(maxsize=None)
-def get_dataset_list():
+
+@save_df(day_lapse_max=30)
```
[the remark this answers is not in the document]
Comment (hadrilec) [reply pynsee/macrodata/get_dataset_list.py:19]: yes indeed, 30 days

File: pynsee/macrodata/_get_dataset_metadata.py [hunk 6]
```diff
@@ -70,18 +72,20 @@ def _get_dataset_metadata(dataset, update=False):
...
-            logger.info(f"Data saved: {file_dataset_metadata}")
+            if not silent:
+                logger.info(f"Data saved: {file_dataset_metadata}")
         else:
             # pickle format depends on python version
             # then read_pickle can fail, if so
             # the file is removed and the function is launched again
             # testing requires multiple python versions
             try:
                 idbank_list_dataset = pd.read_pickle(file_dataset_metadata)
```
[the remark this answers is not in the document]
Comment (hadrilec) [reply pynsee/macrodata/_get_dataset_metadata.py]: save_df is going to be used in this case

File: pynsee/macrodata/_get_dataset_metadata.py [hunk 7]
```diff
@@ -70,18 +72,20 @@ def _get_dataset_metadata(dataset, update=False):
 
             # save data
             idbank_list_dataset.to_pickle(file_dataset_metadata)
```
[the remark this answers is not in the document]
Comment (hadrilec) [reply pynsee/macrodata/_get_dataset_metadata.py]: save_df is going to be used in this case

File: pynsee/utils/save_df.py [hunk 8]
```diff
@@ -44,15 +44,9 @@ def wrapper(*args, **kwargs):
             else:
                 file_name += ".pkl"            
                                     
-            if any([(a == 'update') & (kwargs[a] == True) for a in kwargs.keys()]):
-                update = True
-            else:
-                update = False  
+            update = any((a == 'update') and (kwargs[a] is True) for a in kwargs)
```
Comment (tfardet) [inline pynsee/utils/save_df.py:47]: After suggesting the change to a single liner, I think I know realize that this should be even simpler, shouldn't it?
```suggestion
            update = kwargs.get("update", False)
```

File: pynsee/utils/save_df.py [hunk 9]
```diff
@@ -44,15 +44,9 @@ def wrapper(*args, **kwargs):
...
 
-            if any([(a == 'silent') & (kwargs[a] == True) for a in kwargs.keys()]):
-                silent = True
-            else:
-                silent = False    
+            silent = any((a == 'silent') and (kwargs[a] is True) for a in kwargs)
```
Comment (tfardet) [inline pynsee/utils/save_df.py:49]: If the above is correct, then this one should also be
```suggestion
            silent = kwargs.get("silence", False)
```

File: pynsee/utils/save_df.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+            if os.path.exists(file_name):
+                file_date_last_modif = datetime.datetime.fromtimestamp(
+                    os.path.getmtime(file_name)
+                )
+                try:
+                    # only used for testing purposes
+                    insee_date_time_now = os.environ["insee_date_test"]
+                    insee_date_time_now = datetime.datetime.strptime(
+                        insee_date_time_now, "%Y-%m-%d %H:%M:%S.%f"
+                    )
```
Comment (tfardet) [inline pynsee/utils/save_df.py:60]: what is this?
if it's for testing only, shouldn't it check ``kwargs`` for a testing value or something, this looks quite strange...

**Review by tfardet — CHANGES_REQUESTED** [review]
Noticed another few small issues, otherwise I must say this is really looking good: it's always a good thing when there is a lot of red in a PR, the less code the better :+1:

**Comment by hadrilec** [comment]
Hi @tfardet, I still need some time to fine-tune this PR, I will tell you whenever you can start having a final look

File: pynsee/utils/save_df.py (hunk 10 above)
Comment (hadrilec) [reply pynsee/utils/save_df.py]: the following commit uses kwarg  as advised to test the function retriggering feature if cached data is too old
607d370f9d71764561f8e516803a1ae096aa565f

File: pynsee/geodata/get_geodata.py [hunk 11]
```diff
@@ -28,7 +28,10 @@ def get_geodata(id, update=False, crs="EPSG:3857"):
 
     with warnings.catch_warnings():
         warnings.simplefilter("ignore")
-        df = _get_geodata(id=id, update=update, crs=crs)
+        try:
+            df = _get_geodata(id=id, update=update, crs=crs)
+        except:
+            df = _get_geodata(id=id, update=update, crs=crs)
```
[the remark this answers is not in the document]
Comment (hadrilec) [reply pynsee/geodata/get_geodata.py]: it has been removed with the commit adffae9, but there still might be a connection error when the package is used with onyxia datascience platform

File: pynsee/localdata/get_local_data.py [hunk 12]
```diff
@@ -108,50 +103,27 @@ def get_local_data(
...
-        else:
-            logger.info(
-                "Locally saved data has been used\n"
-                "Set update=True to trigger an update"
+            df = _get_insee_local_onegeo(
+                variables, dataset_version, nivgeo, codegeo
             )
+            
+        except Exception as e:
+            #display(e)
```
Comment (tfardet) [inline pynsee/localdata/get_local_data.py:119]: remove

File: pynsee/macrodata/_get_dataset_dimension.py [hunk 13]
```diff
@@ -26,104 +24,61 @@ def _get_dataset_dimension(dataset, update=False):
...
+        api_url=INSEE_api_link_datastructure_dataset,
+    )
 
-        with open(dataset_dimension_file, "wb") as f:
-            f.write(results.content)
+    # create temporary directory
+    dirpath = _get_temp_dir()
 
-        root = ET.parse(dataset_dimension_file).getroot()
+    dataset_dimension_file = dirpath + "\\dataset_dimension_file"
```
Comment (tfardet) [inline pynsee/macrodata/_get_dataset_dimension.py:35]:
```suggestion
    dataset_dimension_file = os.path.join(dirpath, "dataset_dimension_file")
```

File: pynsee/macrodata/_get_dataset_list_internal.py [hunk 14]
```diff
@@ -2,18 +2,19 @@
...
 
+@save_df(day_lapse_max=90)
 def _get_dataset_list_internal():
 
     zip_file = pkg_resources.resource_stream(__name__, "data/dataset_list_internal.zip")
 
-    insee_folder = _create_insee_folder()
-    dataset_file = insee_folder + "/" + "dataset_list_internal.csv"
+    temp_folder = _get_temp_dir()
+    dataset_file = temp_folder + "/" + "dataset_list_internal.csv"
```
Comment (tfardet) [inline pynsee/macrodata/_get_dataset_list_internal.py:14]: use ``join`` with ``from os.path import join``

File: pynsee/macrodata/_get_dataset_metadata.py [hunk 15]
```diff
@@ -112,3 +52,5 @@ def _get_dataset_metadata(dataset, update=False):
         idbank_list_dataset = idbank_list_dataset.reset_index(drop=True)
 
     return idbank_list_dataset
+        
+
```
Comment (tfardet) [inline pynsee/macrodata/_get_dataset_metadata.py:56]: remove

File: pynsee/macrodata/_get_idbank_internal_data.py [hunk 16]
```diff
@@ -6,59 +6,28 @@
...
+    
     zip_file = pkg_resources.resource_stream(
         __name__, "data/idbank_list_internal.zip"
     )
 
-    if (not os.path.exists(data_final_file)) | update:
-        with zipfile.ZipFile(zip_file, "r") as zip_ref:
-            zip_ref.extractall(insee_folder)
+    temp_folder = _get_temp_dir()
+    data_file = temp_folder + "/" + "idbank_list_internal.csv"
```
Comment (tfardet) [inline pynsee/macrodata/_get_idbank_internal_data.py:24]:
```suggestion
    data_file = os.path.join(temp_folder, "idbank_list_internal.csv")
```

File: pynsee/utils/save_df.py [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+
+def save_df(obj=pd.DataFrame, parquet=True, day_lapse_max=None):
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            
+            data_folder = _create_insee_folder()
+            string_file_arg = [str(kwargs[a]) for a in kwargs.keys() if a not in ['update', 'silent']]
+            string_file_arg += [func.__name__] + [str(a) for a in args]            
+            file_name = data_folder + "/" + _hash(''.join(string_file_arg))
```
Comment (tfardet) [inline pynsee/utils/save_df.py:40]:
```suggestion
            file_name = os.path.join(data_folder, _hash(''.join(string_file_arg)))
```

File: pynsee/utils/save_df.py [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+                            df.to_parquet(file_name)
+                        else:
+                            df.to_csv(file_name, index=False)
+                            
+                    except Exception as e:
+                        warnings.warn(str(e))
+                        warnings.warn(f'Error, file not saved:\n{file_name}\n{df}')
+                        warnings.warn('\n')
+                    
+                    df = obj(df)
```
Comment (tfardet) [inline pynsee/utils/save_df.py:89]: remove and wrap the return instead

File: pynsee/utils/save_df.py [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+                    except Exception as e:
+                        warnings.warn(str(e))
+                                                
+                        kwargs2 = kwargs
+                        kwargs2['update'] = True
+                        
+                        warnings.warn('!!! Unable to load data, function retriggered !!!')
+                        
+                        df = func(*args, **kwargs2)
+                        df = obj(df)
```
Comment (tfardet) [inline pynsee/utils/save_df.py:113]: remove

File: pynsee/utils/save_df.py [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+                        df = obj(df)   
+                    else:
+                        if not silent:
+                            
+                            mdate = insee_date_time_now - datetime.timedelta(days=day_lapse)
+                            
+                            _warning_cached_data(file_name,
+                                                 mdate= mdate,
+                                                 day_lapse=day_lapse)
+                        df = obj(df)
```
Comment (tfardet) [inline pynsee/utils/save_df.py:122]: remove

File: pynsee/utils/save_df.py [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+                        if not silent:
+                            
+                            mdate = insee_date_time_now - datetime.timedelta(days=day_lapse)
+                            
+                            _warning_cached_data(file_name,
+                                                 mdate= mdate,
+                                                 day_lapse=day_lapse)
+                        df = obj(df)            
+            
+            return df
```
Comment (tfardet) [inline pynsee/utils/save_df.py:124]:
```suggestion
            return obj(df)
```

File: tests/macrodata/test_pynsee_macrodata.py [hunk 22]
```diff
@@ -160,18 +155,21 @@ def test_get_dataset_1(self):
...
-            self.assertTrue(len(data1.index) < len(data2.index))
+        # def test_get_dataset_3(self):
+        #     data1 = get_dataset(
+        #         "IPC-2015", filter="M......ENSEMBLE...CVS.2015.")
+            
+        #     data2 = get_dataset("IPC-2015", filter="M......ENSEMBLE...CVS.2015.",
+        #                         includeHistory=True,
+        #                         # updatedAfter="2017-07-11T08:45:00" disabled as API does not return anything
+        #                        )
+        #     self.assertTrue(len(data1.index) < len(data2.index))
```
Comment (tfardet) [inline tests/macrodata/test_pynsee_macrodata.py:172]: uncomment or remove?

**Review by tfardet — CHANGES_REQUESTED** [review]
Thanks a lot for all that work!
I just suggested a few improvements here and there.

**Comment by tfardet** [comment]
woups, [7506899](https://github.com/InseeFrLab/pynsee/pull/196/commits/7506899f8adfb766c764e7338ad24fc89a8c5183) was missing indentation apparently, sorry about that...

File: pynsee/utils/save_df.py (hunk 18 above)
Comment (hadrilec) [reply pynsee/utils/save_df.py]: 4cf3d511b1bd7942b9075ccc463d4664fdd5ff5c should tackle the raised issue

File: pynsee/utils/save_df.py (hunk 19 above)
Comment (hadrilec) [reply pynsee/utils/save_df.py]: 4cf3d511b1bd7942b9075ccc463d4664fdd5ff5c should tackle the raised issue

File: pynsee/utils/save_df.py (hunk 20 above)
Comment (hadrilec) [reply pynsee/utils/save_df.py]: 4cf3d511b1bd7942b9075ccc463d4664fdd5ff5c should tackle the raised issue

File: pynsee/utils/save_df.py (hunk 21 above)
Comment (hadrilec) [reply pynsee/utils/save_df.py]: 4cf3d511b1bd7942b9075ccc463d4664fdd5ff5c should tackle the raised issue

File: tests/macrodata/test_pynsee_macrodata.py (hunk 22 above)
Comment (hadrilec) [reply tests/macrodata/test_pynsee_macrodata.py:172]: once I have time I will include this check back, so I prefer to leave it like this for the moment

File: pynsee/localdata/get_local_data.py [hunk 23]
```diff
@@ -108,50 +103,27 @@ def get_local_data(
         
         dataset_version = _find_latest_local_dataset(dataset_version, variables, nivgeo, geocodes[0], update)
        
-    if (not os.path.exists(file_localdata)) or update:
-
-        list_data_all = []       
+    list_data_all = []       
+
```
[the remark this answers is not in the document]
Comment (hadrilec) [reply pynsee/localdata/get_local_data.py:107]: b27d6751ff1dc1738b522be9dce0680fd43e41d7 should tackle the issue

File: pynsee/macrodata/_get_dataset_list_internal.py (hunk 14 above)
Comment (hadrilec) [reply pynsee/macrodata/_get_dataset_list_internal.py]: b27d6751ff1dc1738b522be9dce0680fd43e41d7 should tackle the issue

File: pynsee/macrodata/_get_dataset_metadata.py (hunk 15 above)
Comment (hadrilec) [reply pynsee/macrodata/_get_dataset_metadata.py]: b27d6751ff1dc1738b522be9dce0680fd43e41d7 should tackle the issue

File: pynsee/localdata/get_local_data.py (hunk 12 above)
Comment (hadrilec) [reply pynsee/localdata/get_local_data.py]: b27d6751ff1dc1738b522be9dce0680fd43e41d7 should tackle the issue

**Review by tfardet — APPROVED** [review]
Looks great, thanks!

**Pull request merged**